src/Common/Auth/OneTimeAuth.php

```diff
@@ -0,0 +1,340 @@
+            }
+        }
+        $rtn['pid'] = $auth['pid'];
+        $rtn['pin'] = $t_info['pin'];
+        $rtn['redirect'] = $redirect ?: $t_info['redirect_url'];
+        $rtn['username'] = $auth['portal_username'];
+        $rtn['login_username'] = $auth['portal_login_username'];
+        $rtn['portal_pwd'] = $auth['portal_pwd'];
+        $rtn['onetime_decrypted'] = $one_time;
+
+        $this->updateOnetime($auth['pid'], $one_time);
+        (new SystemLogger())->debug("Onetime successfully decoded. $one_time");
+
+        return $rtn;
+    }
+
+    /**
+     * Credit to Stephen Neilson
+     *
+     * @param $email
+     * @return bool
+     */
+    private function isValidEmail($email): bool
+    {
+        if (preg_match("/^[_a-z0-9-]+(\.[_a-z0-9-\+]+)*@[a-z0-9-]+(\.[a-z0-9-]+)*(\.[a-z]{2,3})$/i", $email)) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @param       $email
+     * @param       $body
+     * @param array $user
+     * @return string
+     */
+    public function emailNotification($email, $body, array $user = ['fname' => 'Portal', 'lname' => 'Administration']): string
+    {
+        $from_name = ($user['fname'] ?? '') . ' ' . ($user['lname'] ?? '');
+        $mail = new MyMailer();
+        $from_name = text($from_name);
+        $from = $GLOBALS["practice_return_email_path"];
+        $mail->AddReplyTo($from, $from_name);
+        $mail->SetFrom($from, $from);
+        $to = $email;
+        $to_name = $email;
+        $mail->AddAddress($to, $to_name);
+        $subject = xlt("Session Request");
+        $mail->isHTML(true);
+        $mail->Subject = $subject;
+        $mail->Body = $body;
+        if ($mail->Send()) {
+            $status = xlt("Email successfully sent.");
+        } else {
+            $status = xlt("Error: Email failed") . text($mail->ErrorInfo);
+        }
+        return $status;
+    }
+
+    /**
+     * @param $site_addr
+     * @param $token_encrypt
+     * @param $forward
+     * @return string
+     */
+    private function encodeLink($site_addr, $token_encrypt, $pid, $forward = ''): string
+    {
+        $site_id = ($_SESSION['site_id'] ?? null) ?: 'default';
+        $format = "%s&%s";
+        if (stripos($site_addr, "?site") === false) {
+            $format = "%s?%s";
+        }
+        if ($this->scope == 'register') {
+            $encoded_link = sprintf($format, attr($site_addr), http_build_query([
+                'forward_email_verify' => $token_encrypt,
+                'site' => $site_id
+            ]));
+        } elseif ($this->scope == 'reset_password') {
+            $encoded_link = sprintf($format, attr($site_addr), http_build_query([
+                'forward' => $token_encrypt,
+                'site' => $site_id
+            ]));
+        } else {
+            $encoded_link = sprintf($format, attr($site_addr), http_build_query([
+                'id' => $pid,
```

Comment:
Do we need to have the pid in this link?  Not sure why but my gut says we don't want to expose the patient pid as part of the link.

Reply:
no, I'll remove